batch.py

```diff
@@ -757,28 +727,54 @@ def manage_concurrency_for_job(job: hb.job.Job):
         pv_files = []
         for gene in genes_list:
 
+            # check if all of this needs running at all
             # wrapped this with output_path
             pv_file = output_path(f'{celltype}/{gene}_pvals.csv')
-
             # always append the file name
             pv_files.append(pv_file)
-
             # check if running is required
             if to_path(pv_file).exists():
                 logging.info(f'We already ran associations for {gene}!')
                 continue
 
+            # genotypes
+            # if they don't already exist, make plink files
+            plink_file = output_path(f'plink_files/{gene}')
+
+            # if the plink files does not exist, generate them
+            if not to_path(f'{plink_file}.bim').exists():
```

Comment:
I think there's an issue with the way this will run - this section determines which jobs will run, but all checks occur prior to the workflow running - 

i.e. ```if not to_path(f'{plink_file}.bim').exists()```  will initially be false, and a job will be scheduled to make that file for every different cell type. That's not a massive problem if you're running for only one cell type, but it could be a waste on a huge dataset

Reply:
yes so this is exactly what i was worried about, and why I initially had the two completely separate to avoid running this multiple times! 

Any solutions beyond the sink thing? It does seem a bit strange to depend on all of the plink jobs across all genes, but equally it all needs to run, so maybe the order doesn't matter too much?

@illusional @lgruen do you remember any similar issues with Kat's pipeline? It seems like the same discussions would have come up there too?

Reply:
I think having two stages is fine (and avoids the problem that @MattWellie points out). But instead of `*genotype_jobs`, can't `prepare_input_job` simply depend on the single corresponding gene plink result?